src/java.base/share/classes/com/sun/crypto/provider/HkdfKeyDerivation.java

```diff
@@ -0,0 +1,399 @@
+            // SecretKey Objects,
+            // unless we were passed something bogus or an unexportable P11 key
+            try {
+                inputKeyMaterial = consolidateKeyMaterial(ikms);
+                salt = consolidateKeyMaterial(salts);
+            } catch (InvalidKeyException ike) {
+                throw (InvalidAlgorithmParameterException) new InvalidAlgorithmParameterException(
+                    "Issue encountered when combining ikm or salt values into"
+                    + " single keys").initCause(ike);
+            }
+            // set this value in the "if"
+            if ((info = anExtractThenExpand.info()) == null) {
+                info = new byte[0];
+            }
+            length = anExtractThenExpand.length();
+            if (length > (hmacLen * 255)) {
+                throw new InvalidAlgorithmParameterException(
+                    "Requested length exceeds maximum allowed key stream "
+                    + "length");
+            }
+            // perform extract and then expand
+            try {
+                byte[] extractResult = hkdfExtract(inputKeyMaterial, (salt
+                                                                      == null) ? null : salt.getEncoded());
+                pseudoRandomKey = new SecretKeySpec(extractResult, "RAW");
+                return Arrays.copyOf(hkdfExpand(pseudoRandomKey, info, length),
+                                     length);
+            } catch (InvalidKeyException ike) {
+                throw (InvalidAlgorithmParameterException) new InvalidAlgorithmParameterException(
+                    "an HKDF ExtractThenExpand could not be initialized with "
+                    + "the given key or salt material").initCause(ike);
+            } catch (NoSuchAlgorithmException nsae) {
+                // This is bubbling up from the getInstance of the Mac/Hmac.
+                // Since we're defining these values internally, it is unlikely.
+                throw new ProviderException(
+                    "could not instantiate a Mac with the provided algorithm",
+                    nsae);
+            }
+        }
+        throw new InvalidAlgorithmParameterException(
+            "an HKDF could not be initialized with the given KDFParameterSpec");
+    }
+
+    private static boolean isNullOrEmpty(Collection<?> c) {
+        return c == null || c.isEmpty();
+    }
+
+    private SecretKey consolidateKeyMaterial(List<SecretKey> keys)
+        throws InvalidKeyException {
+        if (keys != null && !keys.isEmpty()) {
+            ArrayList<SecretKey> localKeys = new ArrayList<>(keys);
+            if (localKeys.size() == 1) {
+                // return this element
+                SecretKey checkIt = localKeys.get(0);
+                byte[] workItemBytes = CipherCore.getKeyBytes(checkIt);
+                return new SecretKeySpec(workItemBytes, "Generic");
+            } else {
```

Comment:
I think this is less error prone and easier to read than what you have below:
```
                ByteArrayOutputStream os = new ByteArrayOutputStream();
                for (SecretKey workItem : localKeys) {
                    try {
                        os.write(CipherCore.getKeyBytes(workItem));
                    } catch (IOException e) {
                        // won't happen
                    }
                }
                return new SecretKeySpec(os.toByteArray(), "Generic");
```
And if your concerned about the extra copy from `toByteArray()`, you could consider using an internal class `AEADBufferedStream` which extends ByteArrayOutputStream but will return the internal copy to avoid extra mem allocation.

Reply:
+1.

Reply:
resolved